infra/base-images/base-builder/jcc/jcc.go

```diff
@@ -0,0 +1,257 @@
+	srcFile, err := os.Open(srcFilename)
+	if err != nil {
+		return err
+	}
+	srcBytes, err := ioutil.ReadAll(srcFile)
+	if err != nil {
+		return err
+	}
+	src := string(srcBytes)
+	newSrc := ReplaceMissingHeader(src, curHeader, replacementHeader)
+	b := []byte(newSrc)
+	err = ioutil.WriteFile(srcFilename, b, 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func ReplaceMissingHeader(src, curHeader, replacementHeader string) string {
+	re := regexp.MustCompile(`#include ["|<]` + curHeader + `["|>]\n`)
+	replacement := "#include \"" + replacementHeader + "\"\n"
+	return re.ReplaceAllString(src, replacement)
+}
+
+func GetHeaderCorrectedFilename(compilerErr string) (string, string, bool) {
+	re := regexp.MustCompile(`(?P<buggy>[a-z\/\-\_0-9A-z\.]+):.* fatal error: .* file not found`)
+	matches := re.FindStringSubmatch(compilerErr)
+	if len(matches) < 2 {
+		return "", "", false
+	}
+	oldFilename := matches[1]
+	base := filepath.Base(oldFilename)
+	root := filepath.Dir(oldFilename)
+	newFilename := root + "/jcc-corrected-" + base
+	return oldFilename, newFilename, true
+}
+
+func GetHeaderCorrectedCmd(cmd []string, compilerErr string) ([]string, string, error) {
+	oldFilename, newFilename, success := GetHeaderCorrectedFilename(compilerErr)
+	if !success {
+		return cmd, "", errors.New("Couldn't find buggy file")
+	}
+	// Make new cmd.
+	newCmd := make([]string, len(cmd))
+	for i, part := range cmd {
+		newCmd[i] = part
+	}
+	found := false
+	for i, filename := range newCmd {
+		if filename == oldFilename {
+			newCmd[i] = newFilename
+			found = true
+			break
+		}
+	}
+	CopyFile(oldFilename, newFilename)
+	if found {
+		return newCmd, newFilename, nil
+	}
+	return cmd, "", errors.New("Couldn't find file")
+}
+
+func CorrectMissingHeaders(bin string, cmd []string) (bool, error) {
+	_, _, stderr := compile(bin, cmd)
+	cmd, correctedFilename, err := GetHeaderCorrectedCmd(cmd, stderr)
+	if err != nil {
+		return false, err
+	}
+	for i := 0; i < MAX_MISSING_HEADER_FIXES; i++ {
+		fixed, hasBrokenHeaders := TryCompileAndFixHeadersOnce(bin, cmd, correctedFilename)
+		if fixed {
+			return true, nil
+		}
+		if !hasBrokenHeaders {
+			return false, nil
+		}
+	}
+	return false, nil
+}
+
+func compile(bin string, args []string) (int, string, string) {
+	cmd := exec.Command(bin, args...)
+	var outb, errb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	cmd.Run()
+	return cmd.ProcessState.ExitCode(), outb.String(), errb.String()
+}
+
+func TryCompileAndFixHeadersOnce(bin string, cmd []string, filename string) (bool, bool) {
+	retcode, _, err := compile(bin, cmd)
+	if retcode == 0 {
+		return true, false
+	}
+	missingHeader, isMissing := ExtractMissingHeader(err)
+	if !isMissing {
+		return false, false
+	}
+
+	newHeaderPath, found := FindMissingHeader(missingHeader)
+	if !found {
+                return false, true
+	}
+	ReplaceMissingHeaderInFile(filename, missingHeader, newHeaderPath)
+	return false, true
```

Comment:
should this be true, true? 

it's also a bit confusing what these states are. 

Is it possible to simplify this into a single bool value indicating whether or not we should keep trying to compile?